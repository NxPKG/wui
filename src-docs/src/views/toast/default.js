import React from 'react';

import { WuiToast } from '../../../../src/components';

export default () => (
  <div>
    <WuiToast
      title="Example of a good toast"
      onClose={() => window.alert('Dismiss toast')}>
      <p>
        A good toast message is short and to the point. It should very rarely
        include multiple paragraphs.
      </p>
    </WuiToast>
  </div>
);
