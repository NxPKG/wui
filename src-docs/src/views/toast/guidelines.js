import React, { useState } from 'react';

import { GuideRule, GuideRuleExample, GuideRuleTitle } from '../../components';

import {
  WuiText,
  WuiButton,
  WuiSpacer,
  WuiFlexGroup,
  WuiFlexItem,
  WuiToast,
  WuiCallOut,
  WuiModal,
  WuiModalBody,
  WuiModalFooter,
  WuiModalHeader,
  WuiModalHeaderTitle,
  WuiOverlayMask,
  WuiCodeBlock,
  WuiTitle,
} from '../../../../src/components';

export default () => {
  const [isModalVisible, setIsModalVisible] = useState(false);

  const closeModal = () => {
    setIsModalVisible(false);
  };

  const showModal = () => {
    setIsModalVisible(true);
  };

  let modal;

  if (isModalVisible) {
    modal = (
      <WuiOverlayMask>
        <WuiModal onClose={closeModal}>
          <WuiModalHeader>
            <WuiModalHeaderTitle>
              Your visualization has an error
            </WuiModalHeaderTitle>
          </WuiModalHeader>

          <WuiModalBody>
            <WuiCallOut
              title="The maximum bucket size of 200 was exceeded"
              color="danger"
              size="s"
              iconType="alert"
            />
            <WuiSpacer size="s" />
            <WuiCodeBlock>
              {`--- FAKE ERROR ---
An extremely long error trace can exist in a modal so you have time
and space to read it properly. Alternatively just link to a full page.
---
                `}
            </WuiCodeBlock>
          </WuiModalBody>

          <WuiModalFooter>
            <WuiButton onClick={closeModal} fill>
              Close
            </WuiButton>
          </WuiModalFooter>
        </WuiModal>
      </WuiOverlayMask>
    );
  }
  return (
    <>
      <WuiText grow={false} className="guideSection__text">
        <p>
          This page documents patterns for using toasts, short messages that
          appears on the lower right corner and time out after a few seconds.
          They are a popular design choice because they don&apos;t need to fit
          in a layout and don&apos;t disrupt the user.
        </p>
      </WuiText>
      <WuiTitle>
        <h1>Toast types</h1>
      </WuiTitle>

      <WuiSpacer size="xl" />

      <WuiFlexGroup alignItems="center">
        <WuiFlexItem grow={false} style={{ minWidth: 120 }}>
          <WuiToast
            style={{ width: 300 }}
            title="Your report is complete"
            color="success"
          />
        </WuiFlexItem>

        <WuiFlexItem>
          <WuiText className="guideSection__text">
            <h3>Success toasts indicate that everything worked out</h3>
            <p>They are the most-commonly used toasts.</p>
          </WuiText>
        </WuiFlexItem>
      </WuiFlexGroup>

      <WuiSpacer />

      <WuiFlexGroup alignItems="center">
        <WuiFlexItem grow={false} style={{ minWidth: 120 }}>
          <WuiToast
            style={{ width: 300 }}
            title="Node 726 is having trouble"
            color="warning"
          />
        </WuiFlexItem>

        <WuiFlexItem>
          <WuiText className="guideSection__text">
            <h3>Warning toasts direct user attention to a potential problem</h3>
            <p>
              These toasts work well in monitoring apps when something
              significant requires action.
            </p>
          </WuiText>
        </WuiFlexItem>
      </WuiFlexGroup>

      <WuiSpacer />

      <WuiFlexGroup alignItems="center">
        <WuiFlexItem grow={false} style={{ minWidth: 120 }}>
          <WuiToast
            style={{ width: 300 }}
            title="Search failed.  Check your Wazuh connection."
            color="danger"
          />
        </WuiFlexItem>

        <WuiFlexItem>
          <WuiText className="guideSection__text">
            <h3>Error toasts report a problem</h3>
            <p>
              An error toast might let users know an action didn&apos;t complete
              or that a form has errors.
            </p>
          </WuiText>
        </WuiFlexItem>
      </WuiFlexGroup>

      <WuiSpacer />

      <WuiFlexGroup alignItems="center">
        <WuiFlexItem grow={false} style={{ minWidth: 120 }}>
          <div>
            <WuiToast
              style={{ width: 300 }}
              title="Please wait while your report is created"
              color="primary"
            />
          </div>
        </WuiFlexItem>

        <WuiFlexItem>
          <WuiText className="guideSection__text">
            <h3>Info toasts relay neutral information</h3>
            <p>
              The default toast, an info toast might notify users about an
              ongoing action.
            </p>
          </WuiText>
        </WuiFlexItem>
      </WuiFlexGroup>

      <WuiSpacer />

      <GuideRuleTitle>Use a toast for a timely message</GuideRuleTitle>

      <GuideRule
        description="Toasts are appropriate for short feedback related to a user action.
          A toast should contain a message about a current action, not a historical action.">
        <GuideRuleExample
          type="do"
          panel={false}
          frame
          text="Do. Use a toast for a brief message about the current action.">
          <WuiToast
            style={{ maxWidth: 300 }}
            title="Your folder was moved"
            color="success"
          />
        </GuideRuleExample>

        <GuideRuleExample
          panel={false}
          type="dont"
          frame
          text="Don't greet users with a toast when they open a page.">
          <div style={{ textAlign: 'center' }}>
            <WuiToast
              style={{ maxWidth: 300 }}
              title="Haven't seen you in a while"
              color="primary"
            />
          </div>
        </GuideRuleExample>
      </GuideRule>

      <GuideRuleTitle>
        Most often, it&apos;s a single line of text
      </GuideRuleTitle>

      <GuideRule
        description="By default, a toast stays on the screen 10 seconds.
          Users should be able read the message in 6 to 7 seconds.
          The message should get straight to the point and rarely include more than one line.

          ">
        <GuideRuleExample
          panel={false}
          frame
          type="do"
          text="Do. A single line of text is readable at a glance.">
          <div style={{ textAlign: 'center' }}>
            <WuiToast
              style={{ maxWidth: 300 }}
              title="Check your form for errors"
              color="danger"
            />
          </div>
        </GuideRuleExample>

        <GuideRuleExample
          type="dont"
          panel={false}
          frame
          text="Don't cram a lot of detail into a toast.
            These errors should persist in callouts and validations on the form.
            They don't need to be spelled out in the toast.">
          <div>
            <WuiToast
              style={{ maxWidth: 300 }}
              title="Your form has errors"
              color="danger">
              <WuiText className="guideSection__text">
                <ul>
                  <li>Username is a required field.</li>
                  <li>Password must be at least 6 characters long.</li>
                  <li>Email is a required field.</li>
                </ul>
              </WuiText>
            </WuiToast>
          </div>
        </GuideRuleExample>
      </GuideRule>

      <GuideRuleTitle>
        Toasts should only contain a single action
      </GuideRuleTitle>

      <GuideRule
        description="A toast can have a single action, styled as a standard button.
          If more actions are needed, or if the action is important enough to
          interrupt the user, use a modal instead.">
        <GuideRuleExample
          panel={false}
          type="do"
          frame
          text="Do. Use only one action per toast and favor a one-word label.
              Align actions to the right, which follows our button guidelines for
              usage within restricted width containers.">
          <WuiToast
            style={{ maxWidth: 300 }}
            color="success"
            title="Your report is complete">
            <div style={{ textAlign: 'right' }}>
              <WuiButton size="s">Download</WuiButton>
            </div>
          </WuiToast>
        </GuideRuleExample>

        <GuideRuleExample
          type="dont"
          panel={false}
          frame
          text="Don't use multiple actions. Don't align buttons in toasts to the left.
              This message is better in a confirmation modal.">
          <WuiToast
            style={{ maxWidth: 300 }}
            title="All messages will be deleted"
            color="danger">
            <WuiFlexGroup justifyContent="flexEnd" gutterSize="s">
              <WuiFlexItem grow={false}>
                <WuiButton size="s">Cancel</WuiButton>
              </WuiFlexItem>
              <WuiFlexItem grow={false}>
                <WuiButton size="s" color="danger">
                  Delete
                </WuiButton>
              </WuiFlexItem>
            </WuiFlexGroup>
          </WuiToast>
        </GuideRuleExample>
      </GuideRule>

      <WuiSpacer size="l" />

      <GuideRuleTitle>Icons should emphasize actions</GuideRuleTitle>

      <GuideRule description="An icon on the left of the message can help define the message type.">
        <GuideRuleExample
          panel={false}
          type="do"
          frame
          text="Do. The check icon reinforces that the action succeeded.
                The alert icon helps users understand the message is an error.">
          <div>
            <WuiToast
              style={{ maxWidth: 300 }}
              title="Your dashboard was updated"
              iconType="check"
              color="success"
            />

            <WuiSpacer />

            <WuiToast
              style={{ maxWidth: 300 }}
              title="A dashboard named 'MyDashboard' already exists"
              iconType="alert"
              color="danger"
            />
          </div>
        </GuideRuleExample>

        <GuideRuleExample
          type="dont"
          panel={false}
          frame
          text="Don't use icons that are hard to understand. They distract from the message.">
          <WuiToast
            color="primary"
            style={{ maxWidth: 300 }}
            title="Message sent"
            iconType="help"
          />
        </GuideRuleExample>
      </GuideRule>

      <GuideRuleTitle>Display one toast at a time</GuideRuleTitle>

      <GuideRule
        description="Users should be able to take
          in all the details from one toast before the next one arrives.">
        <GuideRuleExample
          panel={false}
          type="do"
          frame
          text="Do. Display one toast at a time.">
          <WuiToast
            style={{ maxWidth: 300 }}
            color="primary"
            title="3 new messages"
          />
        </GuideRuleExample>

        <GuideRuleExample
          type="dont"
          panel={false}
          frame
          text="Don't stack toasts.">
          <div>
            <WuiToast
              style={{ maxWidth: 300 }}
              color="danger"
              title="There was a problem with your node">
              <div style={{ textAlign: 'right' }}>
                <WuiButton size="s">Learn more</WuiButton>
              </div>
            </WuiToast>
            <WuiSpacer />

            <WuiToast
              color="primary"
              style={{ maxWidth: 300 }}
              title="3 new messages"
            />
          </div>
        </GuideRuleExample>
      </GuideRule>

      <GuideRuleTitle>Keep messages as short as possible</GuideRuleTitle>

      <GuideRule
        description="For common actions such as create, add, delete, remove, and save,
          include the object type, the object name if available, and the past tense of the action.
          ">
        <GuideRuleExample
          panel={false}
          type="do"
          frame
          text="Do. Include the object name if it's not too long.
            Use single quotation marks around the object name if it helps clarify meaning.">
          <div>
            <WuiToast
              color="success"
              style={{ maxWidth: 300 }}
              title="User 'Casey Smith' was added"
            />
          </div>
        </GuideRuleExample>

        <GuideRuleExample
          type="dont"
          panel={false}
          frame
          text='Don&apos;t use the generic "Your object."'>
          <WuiToast
            color="success"
            style={{ maxWidth: 300 }}
            title="Your object has been saved"
          />
        </GuideRuleExample>
      </GuideRule>

      <GuideRule description="Don't include the word &quot;successfully.&quot; It's implied.">
        <GuideRuleExample
          panel={false}
          frame
          type="do"
          text="Do. Use this format for a success message.">
          <WuiToast
            color="success"
            style={{ maxWidth: 300 }}
            title="Dashboard 'My_dashboard_with_a_very_long_name' was saved"
          />
        </GuideRuleExample>

        <GuideRuleExample
          type="dont"
          panel={false}
          frame
          text='Don&apos;t include "successfully."'>
          <WuiToast
            color="success"
            style={{ maxWidth: 300 }}
            title="Dashboard 'My_dashboard' was successfully saved"
          />
        </GuideRuleExample>
      </GuideRule>

      <GuideRule description="For a message about multiple objects, include the object count, but not the names of the objects.">
        <GuideRuleExample
          panel={false}
          type="do"
          frame
          text="Do. Include the object count.">
          <WuiToast
            color="success"
            style={{ maxWidth: 300 }}
            title="4 visualizations were deleted"
          />
        </GuideRuleExample>
        <GuideRuleExample
          panel={false}
          type="dont"
          frame
          text="Don't overwhelm the user by listing the names of all the objects.">
          <WuiToast
            color="success"
            style={{ maxWidth: 300 }}
            title="Visualization 1, Visualization 2, Visualization 3, and Visualization 4 were deleted"
          />
        </GuideRuleExample>
      </GuideRule>

      <GuideRuleTitle>
        Use call-to-action buttons when the content needs more room
      </GuideRuleTitle>

      <GuideRule
        description="Occassionally the content of a toast is too involved to fit into the constrained space of a toast.
          This is common in long error messages. In these cases use the toast to deliver the summary of the
          information and use a button to provide a call-to-action for the full message.">
        <GuideRuleExample
          panel={false}
          type="do"
          frame
          text="Use the toast message to provide a summary and a button to link to the full content">
          <WuiToast
            style={{ maxWidth: 300 }}
            color="danger"
            title="Your visualization has an error">
            <p>The maximum bucket size of 200 was exceeded.</p>
            <div style={{ textAlign: 'right' }}>
              <WuiButton size="s" color="danger" onClick={showModal}>
                See the full error
              </WuiButton>
              {modal}
            </div>
          </WuiToast>
        </GuideRuleExample>

        <GuideRuleExample
          type="dont"
          panel={false}
          frame
          text="Don't cram a lot of content into the small space of a toast.">
          <WuiToast
            style={{ maxWidth: 300 }}
            color="danger"
            title="Your visualization has an error">
            <WuiCallOut
              title="The maximum bucket size of 200 was exceeded"
              color="danger">
              <p>An extremely long error trace.</p>
            </WuiCallOut>
          </WuiToast>
        </GuideRuleExample>
      </GuideRule>
    </>
  );
};
