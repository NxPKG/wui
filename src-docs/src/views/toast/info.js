import React from 'react';

import { WuiToast } from '../../../../src/components';

export default () => (
  <WuiToast
    title="Icons should be rare"
    type="info"
    onClose={() => window.alert('Dismiss toast')}>
    <p>
      Icons should be used rarely. They are good for warnings, but when paired
      with long titles they look out of place.
    </p>
  </WuiToast>
);
