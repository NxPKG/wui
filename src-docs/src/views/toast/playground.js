import { PropTypes } from 'react-view';
import { WuiToast } from '../../../../src/components/';
import {
  propUtilityForPlayground,
  iconValidator,
  createOptionalEnum,
  dummyFunction,
  simulateFunction,
} from '../../services/playground';

export default () => {
  const docgenInfo = Array.isArray(WuiToast.__docgenInfo)
    ? WuiToast.__docgenInfo[0]
    : WuiToast.__docgenInfo;
  const propsToUse = propUtilityForPlayground(docgenInfo.props);

  propsToUse.iconType = iconValidator(propsToUse.iconType);

  propsToUse.title = {
    ...propsToUse.title,
    type: PropTypes.String,
    value: 'Toast content',
  };

  propsToUse.color = createOptionalEnum(propsToUse.color);

  propsToUse.onClose = simulateFunction(propsToUse.onClose);

  return {
    config: {
      componentName: 'WuiToast',
      props: propsToUse,
      scope: {
        WuiToast,
      },
      imports: {
        '@wazuh/wui': {
          named: ['WuiToast'],
        },
      },
      customProps: {
        onClose: dummyFunction,
      },
    },
  };
};
