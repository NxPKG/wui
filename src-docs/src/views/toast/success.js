import React from 'react';

import {
  WuiButton,
  WuiLink,
  WuiToast,
  WuiFlexGroup,
  WuiFlexItem,
} from '../../../../src/components';

export default () => (
  <WuiToast
    title="Created report for dashboard 'Tuba Sales by Region'"
    color="success"
    iconType="check">
    <p>
      While the layout will adjust properly for wrapping titles, they do not
      look particularly good. Similarily, do not use a whole lot of text in your
      body either. At a certain point people will not have enough time to read
      these things. Like, you probably are not even reading this now.
    </p>

    <p>
      And some other stuff on another line, just for kicks. And{' '}
      <WuiLink href="#">here&rsquo;s a link</WuiLink>.
    </p>

    <WuiFlexGroup justifyContent="flexEnd" gutterSize="s">
      <WuiFlexItem grow={false}>
        <WuiButton size="s">Download report</WuiButton>
      </WuiFlexItem>
    </WuiFlexGroup>
  </WuiToast>
);
