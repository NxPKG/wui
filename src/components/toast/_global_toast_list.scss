/**
 * 1. Allow list to expand as items are added, but cap it at the screen height.
 * 2. Allow some padding for shadow
 */
.wuiGlobalToastList {
  @include wuiScrollBar;
  display: flex;
  flex-direction: column;
  align-items: stretch;
  position: fixed;
  z-index: $wuiZToastList;
  bottom: 0;
  width: $wuiToastWidth + ($wuiSize * 5); /* 2 */
  max-height: 100vh; /* 1 */
  overflow-y: auto;

  // Hide the scrollbar entirely
  // sass-lint:disable-block no-misspelled-properties
  scrollbar-width: none;

  // sass-lint:disable-block no-vendor-prefixes
  &::-webkit-scrollbar {
    width: 0;
    height: 0;
  }

  // The top and bottom padding give height to the list creating a dead-zone effect
  // when there's no toasts in the list, meaning you can't click anything beneath it.
  // Only add the padding if there's content.
  &:not(:empty) {
    padding: $wuiSize;
  }
}

.wuiGlobalToastList--right:not(:empty) {
  right: 0;
  padding-left: $wuiSize * 4; /* 2 */
}

.wuiGlobalToastList--left:not(:empty) {
  left: 0;
  padding-right: $wuiSize * 4; /* 2 */
}

.wuiGlobalToastListItem {
  margin-bottom: $wuiSize;
  animation: $wuiAnimSpeedNormal wuiShowToast $wuiAnimSlightResistance;
  opacity: 1;

  /**
   * 1. justify-content: flex-end interferes with overflowing content, so we'll use this to push
   *    items to the bottom instead.
   */
  &:first-child {
    margin-top: auto; /* 1 */
  }

  &:last-child {
    margin-bottom: 0;
  }

  &.wuiGlobalToastListItem-isDismissed {
    transition: opacity $wuiAnimSpeedNormal;
    opacity: 0;
  }
}

@keyframes wuiShowToast {
  from {
    transform: translateY($wuiSizeL) scale(.9);
    opacity: 0;
  }

  to {
    transform: translateY(0) scale(1);
    opacity: 1;
  }
}

@include wuiBreakpoint('xs', 's') {
  /**
   * 1. Mobile we make these 100%. Matching change happens on the item as well.
   */
  .wuiGlobalToastList:not(:empty) {
    left: 0;
    padding-left: $wuiSize;
    padding-right: $wuiSize;
    width: 100%; /* 1 */
  }
}
