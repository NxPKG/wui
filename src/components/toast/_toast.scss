.wuiToast {
  border: $wuiBorderThin;
  // This overwrites some of the border above
  // sass-lint:disable-block mixins-before-declarations
  @include wuiBottomShadowLarge($adjustBorders: true);

  position: relative;
  padding: $wuiSize;
  background-color: $wuiColorEmptyShade;
  width: 100%;

  &:hover .wuiToast__closeButton,
  &:focus .wuiToast__closeButton {
    opacity: 1;
  }
}

/**
  * 1. Fit button to icon.
  */
.wuiToast__closeButton {
  position: absolute;
  top: $wuiSize;
  right: $wuiSize;
  line-height: 0; /* 1 */
  appearance: none;
  opacity: 0;
  transition: opacity $wuiAnimSpeedFast $wuiAnimSlightResistance;

  svg {
    fill: tintOrShade($wuiTitleColor, 50%, 70%);
  }

  &:hover {
    svg {
      fill: $wuiTitleColor;
    }
  }

  &:focus {
    background-color: $wuiFocusBackgroundColor;
    opacity: 1;

    svg {
      fill: $wuiColorPrimary;
    }
  }
}

// Modifier naming and colors.
$wuiToastTypes: (
  primary: makeGraphicContrastColor($wuiColorPrimary, $wuiColorEmptyShade),
  success: makeGraphicContrastColor($wuiColorSecondary, $wuiColorEmptyShade),
  warning: makeGraphicContrastColor($wuiColorWarning, $wuiColorEmptyShade),
  danger: makeGraphicContrastColor($wuiColorDanger, $wuiColorEmptyShade),
);

// Create button modifiers based upon the map.
@each $name, $color in $wuiToastTypes {
  .wuiToast--#{$name} {
    border-top: 2px solid $color;
  }
}

/**
 * 1. Align icon with first line of title text if it wraps.
 * 2. Apply margin to all but last item in the flex.
 * 3. Account for close button.
 */
.wuiToastHeader {
  padding-right: $wuiSizeL; /* 3 */
  display: flex;
  align-items: baseline; /* 1 */

  > * + * {
    margin-left: $wuiSizeS; /* 2 */
  }
}

/**
  * 1. Vertically center icon with first line of title.
  */
.wuiToastHeader__icon {
  flex: 0 0 auto;
  fill: $wuiTitleColor;
  transform: translateY(2px); /* 1 */
}

.wuiToastHeader__title {
  @include wuiTitle('xs');
  @include wuiTextBreakWord;
  font-weight: $wuiFontWeightLight;
}

.wuiToastHeader--withBody {
  margin-bottom: $wuiSizeS;
}

/**
 * 1. Prevent long lines from overflowing.
 */
.wuiToastBody {
  word-wrap: break-word; /* 1 */
}
