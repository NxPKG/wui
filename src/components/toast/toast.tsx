/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, {
  FunctionComponent,
  HTMLAttributes,
  ReactElement,
  ReactNode,
} from 'react';
import classNames from 'classnames';

import { CommonProps, keysOf } from '../common';
import { WuiScreenReaderOnly } from '../accessibility';
import { WuiI18n } from '../i18n';

import { IconType, WuiIcon } from '../icon';

import { WuiText } from '../text';

type ToastColor = 'primary' | 'success' | 'warning' | 'danger';

const colorToClassNameMap: { [color in ToastColor]: string } = {
  primary: 'wuiToast--primary',
  success: 'wuiToast--success',
  warning: 'wuiToast--warning',
  danger: 'wuiToast--danger',
};

export const COLORS = keysOf(colorToClassNameMap);

export interface WuiToastProps
  extends CommonProps,
    Omit<HTMLAttributes<HTMLDivElement>, 'title'> {
  title?: ReactNode;
  color?: ToastColor;
  iconType?: IconType;
  onClose?: () => void;
}

export const WuiToast: FunctionComponent<WuiToastProps> = ({
  title,
  color,
  iconType,
  onClose,
  children,
  className,
  ...rest
}) => {
  const classes = classNames(
    'wuiToast',
    color ? colorToClassNameMap[color] : null,
    className
  );
  const headerClasses = classNames('wuiToastHeader', {
    'wuiToastHeader--withBody': children,
  });

  let headerIcon: ReactElement;

  if (iconType) {
    headerIcon = (
      <WuiIcon
        className="wuiToastHeader__icon"
        type={iconType}
        size="m"
        aria-hidden="true"
      />
    );
  }

  let closeButton;

  if (onClose) {
    closeButton = (
      <WuiI18n token="wuiToast.dismissToast" default="Dismiss toast">
        {(dismissToast: string) => (
          <button
            type="button"
            className="wuiToast__closeButton"
            aria-label={dismissToast}
            onClick={onClose}
            data-test-subj="toastCloseButton">
            <WuiIcon type="cross" size="m" aria-hidden="true" />
          </button>
        )}
      </WuiI18n>
    );
  }

  let optionalBody;

  if (children) {
    optionalBody = (
      <WuiText size="s" className="wuiToastBody">
        {children}
      </WuiText>
    );
  }

  return (
    <div className={classes} {...rest}>
      <WuiScreenReaderOnly>
        <p>
          <WuiI18n
            token="wuiToast.newNotification"
            default="A new notification appears"
          />
        </p>
      </WuiScreenReaderOnly>

      <WuiI18n token="wuiToast.notification" default="Notification">
        {(notification: string) => (
          <div
            className={headerClasses}
            aria-label={notification}
            data-test-subj="wuiToastHeader">
            {headerIcon}

            <span className="wuiToastHeader__title">{title}</span>
          </div>
        )}
      </WuiI18n>

      {closeButton}
      {optionalBody}
    </div>
  );
};
