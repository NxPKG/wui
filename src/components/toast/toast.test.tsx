/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render, mount } from 'enzyme';
import { findTestSubject, requiredProps } from '../../test';

import { COLORS, WuiToast } from './toast';

describe('WuiToast', () => {
  test('is rendered', () => {
    const component = render(
      <WuiToast {...requiredProps} title="test title">
        <p>Hi</p>
      </WuiToast>
    );

    expect(component).toMatchSnapshot();
  });

  describe('Props', () => {
    describe('title', () => {
      test('is rendered', () => {
        const component = <WuiToast title="toast title" />;
        expect(mount(component)).toMatchSnapshot();
      });
    });

    describe('color', () => {
      COLORS.forEach(color => {
        test(`${color} is rendered`, () => {
          const component = <WuiToast color={color} title="test title" />;
          expect(mount(component)).toMatchSnapshot();
        });
      });
    });

    describe('iconType', () => {
      test('is rendered', () => {
        const component = <WuiToast iconType="user" title="test title" />;
        expect(mount(component)).toMatchSnapshot();
      });
    });

    describe('onClose', () => {
      test('is called when the close button is clicked', () => {
        const onCloseHandler = jest.fn();

        const component = mount(
          <WuiToast onClose={onCloseHandler} title="test title" />
        );
        const closeButton = findTestSubject(component, 'toastCloseButton');
        closeButton.simulate('click');

        expect(onCloseHandler).toBeCalledTimes(1);
      });
    });
  });
});
