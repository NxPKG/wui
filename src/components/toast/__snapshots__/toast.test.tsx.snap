// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WuiToast Props color danger is rendered 1`] = `
<WuiToast
  color="danger"
  title="test title"
>
  <div
    className="wuiToast wuiToast--danger"
  >
    <WuiScreenReaderOnly>
      <p
        className="wuiScreenReaderOnly"
      >
        <WuiI18n
          default="A new notification appears"
          token="wuiToast.newNotification"
        >
          A new notification appears
        </WuiI18n>
      </p>
    </WuiScreenReaderOnly>
    <WuiI18n
      default="Notification"
      token="wuiToast.notification"
    >
      <div
        aria-label="Notification"
        className="wuiToastHeader"
        data-test-subj="wuiToastHeader"
      >
        <span
          className="wuiToastHeader__title"
        >
          test title
        </span>
      </div>
    </WuiI18n>
  </div>
</WuiToast>
`;

exports[`WuiToast Props color primary is rendered 1`] = `
<WuiToast
  color="primary"
  title="test title"
>
  <div
    className="wuiToast wuiToast--primary"
  >
    <WuiScreenReaderOnly>
      <p
        className="wuiScreenReaderOnly"
      >
        <WuiI18n
          default="A new notification appears"
          token="wuiToast.newNotification"
        >
          A new notification appears
        </WuiI18n>
      </p>
    </WuiScreenReaderOnly>
    <WuiI18n
      default="Notification"
      token="wuiToast.notification"
    >
      <div
        aria-label="Notification"
        className="wuiToastHeader"
        data-test-subj="wuiToastHeader"
      >
        <span
          className="wuiToastHeader__title"
        >
          test title
        </span>
      </div>
    </WuiI18n>
  </div>
</WuiToast>
`;

exports[`WuiToast Props color success is rendered 1`] = `
<WuiToast
  color="success"
  title="test title"
>
  <div
    className="wuiToast wuiToast--success"
  >
    <WuiScreenReaderOnly>
      <p
        className="wuiScreenReaderOnly"
      >
        <WuiI18n
          default="A new notification appears"
          token="wuiToast.newNotification"
        >
          A new notification appears
        </WuiI18n>
      </p>
    </WuiScreenReaderOnly>
    <WuiI18n
      default="Notification"
      token="wuiToast.notification"
    >
      <div
        aria-label="Notification"
        className="wuiToastHeader"
        data-test-subj="wuiToastHeader"
      >
        <span
          className="wuiToastHeader__title"
        >
          test title
        </span>
      </div>
    </WuiI18n>
  </div>
</WuiToast>
`;

exports[`WuiToast Props color warning is rendered 1`] = `
<WuiToast
  color="warning"
  title="test title"
>
  <div
    className="wuiToast wuiToast--warning"
  >
    <WuiScreenReaderOnly>
      <p
        className="wuiScreenReaderOnly"
      >
        <WuiI18n
          default="A new notification appears"
          token="wuiToast.newNotification"
        >
          A new notification appears
        </WuiI18n>
      </p>
    </WuiScreenReaderOnly>
    <WuiI18n
      default="Notification"
      token="wuiToast.notification"
    >
      <div
        aria-label="Notification"
        className="wuiToastHeader"
        data-test-subj="wuiToastHeader"
      >
        <span
          className="wuiToastHeader__title"
        >
          test title
        </span>
      </div>
    </WuiI18n>
  </div>
</WuiToast>
`;

exports[`WuiToast Props iconType is rendered 1`] = `
<WuiToast
  iconType="user"
  title="test title"
>
  <div
    className="wuiToast"
  >
    <WuiScreenReaderOnly>
      <p
        className="wuiScreenReaderOnly"
      >
        <WuiI18n
          default="A new notification appears"
          token="wuiToast.newNotification"
        >
          A new notification appears
        </WuiI18n>
      </p>
    </WuiScreenReaderOnly>
    <WuiI18n
      default="Notification"
      token="wuiToast.notification"
    >
      <div
        aria-label="Notification"
        className="wuiToastHeader"
        data-test-subj="wuiToastHeader"
      >
        <WuiIcon
          aria-hidden="true"
          className="wuiToastHeader__icon"
          size="m"
          type="user"
        >
          <div
            aria-hidden="true"
            className="wuiToastHeader__icon"
            data-wuiicon-type="user"
            size="m"
          />
        </WuiIcon>
        <span
          className="wuiToastHeader__title"
        >
          test title
        </span>
      </div>
    </WuiI18n>
  </div>
</WuiToast>
`;

exports[`WuiToast Props title is rendered 1`] = `
<WuiToast
  title="toast title"
>
  <div
    className="wuiToast"
  >
    <WuiScreenReaderOnly>
      <p
        className="wuiScreenReaderOnly"
      >
        <WuiI18n
          default="A new notification appears"
          token="wuiToast.newNotification"
        >
          A new notification appears
        </WuiI18n>
      </p>
    </WuiScreenReaderOnly>
    <WuiI18n
      default="Notification"
      token="wuiToast.notification"
    >
      <div
        aria-label="Notification"
        className="wuiToastHeader"
        data-test-subj="wuiToastHeader"
      >
        <span
          className="wuiToastHeader__title"
        >
          toast title
        </span>
      </div>
    </WuiI18n>
  </div>
</WuiToast>
`;

exports[`WuiToast is rendered 1`] = `
<div
  aria-label="aria-label"
  class="wuiToast testClass1 testClass2"
  data-test-subj="test subject string"
>
  <p
    class="wuiScreenReaderOnly"
  >
    A new notification appears
  </p>
  <div
    aria-label="Notification"
    class="wuiToastHeader wuiToastHeader--withBody"
    data-test-subj="wuiToastHeader"
  >
    <span
      class="wuiToastHeader__title"
    >
      test title
    </span>
  </div>
  <div
    class="wuiText wuiText--small wuiToastBody"
  >
    <p>
      Hi
    </p>
  </div>
</div>
`;
